docs/expressions/functional-operations.md

```diff
@@ -0,0 +1,56 @@
+# Expressions: Functional Operations
+
+Functional operations refer to a style of programming that emphasizes the use of functions as first-class citizens.
+
+## Functional Expression
+
+A functional expression is a way of creating an anonymous function, which is a function without a name, that can be immediately invoked or passed as an argument to another function.
+
+For example:
+
+```
+$a = $(123);
+```
+
+This is the same as:
+
+```
+$a = static function(): int { return 123; };
+```
+
+Another example:
+
+```
+$a = $($this);
+```
+
+Is the same as:
+
+``` 
+$a = function(): self { return $this; };
+```
+
+In both examples, you can see that we need much less code to write the functional expression.
```

Comment:
here's a more clear description/examples ( the links to exit and throw must be updated )

---

A functional expression is a way of create an anonymous function ( also know as [`Closure`](https://www.php.net/manual/en/class.closure.php) ) from any expression.

The resulting `Closure` will have the type of `Closure<(), T>` where `T` is the resulted value type of the expression.

Example:

```
$a = 1;
$b = 2;
$c = 3;

$closure = $($a + $b + $c);
// same as:
$closure = function() use($a, $b, $c): int {
  return $a + $b + $c;
};
```

If the inner expression is unreadable, `T` would be of type `void`.

```
$closure = $($this->counter++);
// same as:
$closure = function(): void {
  $this->counter++;
};
```

If the inner expression is an [exit](#link), or [throw](#link) operation, `T` would be of type `never`.

```
$closure = $(throw new Exception('...'));
// same as:
$closure = static function(): never {
  throw new Exception('...')
};
```

In both examples, you can see that we need much less code to write the functional expression.